pmlc/rt/level_zero/level_zero_invocation.cc

```diff
@@ -0,0 +1,191 @@
+// Copyright 2020 Intel Corporation
+#include "pmlc/rt/level_zero/level_zero_invocation.h"
+
+#include <algorithm>
+#include <chrono>
+#include <string>
+#include <utility>
+
+#include "pmlc/util/logging.h"
+
+namespace pmlc::rt::level_zero {
+
+void LevelZeroMemory::enqueueRead(ze_command_list_handle_t list, void *dst,
+                                  std::vector<ze_event_handle_t> &dependencies,
+                                  ze_event_handle_t &resultE) {
+  lzu::append_memory_copy(list, dst, buffer, bytes, resultE,
+                          dependencies.size(), dependencies.data());
+}
+
+void LevelZeroMemory::enqueueWrite(ze_command_list_handle_t list, void *src,
+                                   std::vector<ze_event_handle_t> &dependencies,
+                                   ze_event_handle_t &resultE) {
+  lzu::append_memory_copy(list, buffer, src, bytes, resultE,
+                          dependencies.size(), dependencies.data());
+}
+
+LevelZeroKernel::LevelZeroKernel(ze_module_handle_t module, std::string name)
+    : module(module), name(name) {
+  kernel = lzu::create_function(module, /*flag*/ 0, name);
+}
+
+LevelZeroKernel::~LevelZeroKernel() {
+  lzu::destroy_function(kernel);
+  lzu::destroy_module(module);
+}
+
+void LevelZeroKernel::addDependency(LevelZeroEvent *event) {
+  dependencies.push_back(event->getEvent());
+}
+
+void LevelZeroKernel::setArg(unsigned idx, LevelZeroMemory *memory) {
+  void *buf = reinterpret_cast<int64_t *>(memory->getBuffer());
+  lzu::set_argument_value(kernel, idx, sizeof(buf), &buf);
+}
+
+void LevelZeroKernel::enqueue(ze_command_list_handle_t list,
+                              ze_group_count_t gws, ze_group_count_t lws,
+                              ze_event_handle_t &resultE) {
+  // The usage of gws and lws will depend on their meaning.
+  // "gws / lws" in some cases.
+  lzu::append_launch_function(list, kernel, &gws, resultE, dependencies.size(),
+                              dependencies.data());
+  dependencies.clear();
+}
+
+LevelZeroEvent::LevelZeroEvent(ze_event_handle_t event,
+                               LevelZeroActionKind kind, std::string name)
+    : event(event), kind(kind), name(std::move(name)) {}
+
+void LevelZeroEvent::wait(const std::vector<LevelZeroEvent *> &events) {
+  std::vector<ze_event_handle_t> zeEvents;
+  std::transform(events.begin(), events.end(), std::back_inserter(zeEvents),
+                 [](const LevelZeroEvent *event) { return event->getEvent(); });
+  for (auto e : zeEvents) {
+    zeEventHostSynchronize(e, UINT64_MAX);
+  }
+}
+
+LevelZeroInvocation::LevelZeroInvocation(LevelZeroDevice *device)
+    : device{device->shared_from_this()}, queueUser(device->getQueue(p)) {
```

Comment:
Is the `p` in ` queueUser(device->getQueue(p)` a global vaiable? Can we changed to some default settings as in line 76-78?

Reply:
ok, it is a private member now.

Reply:
Yes, the queue property is not really used now, as we use command queue and command list in L0, we may do complex check when more features are opened.